# ML Fundamentals #5: Support Vector Machine (SVM)

## Heart Disease Prediction using Support Vector Machine

This project implements a Support Vector Machine classifier to predict the presence of heart disease using patient clinical and diagnostic attributes.

The notebook covers:

- Data Loading
- Dataset Understanding
- Exploratory Data Analysis
- Data Preprocessing
- Feature Scaling
- SVM Model Training
- Prediction
- Model Evaluation
- Kernel Comparison
- Decision Boundary Analysis
- Hyperparameter Tuning
- Model Serialization

In [2]:
# Import required libraries for analysis, visualization, preprocessing, SVM modeling, and evaluation

import os
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
    classification_report,
    roc_curve
)

import joblib

In [3]:
# Create directories for models, figures, and tables

os.makedirs("models", exist_ok=True)
os.makedirs("outputs/figures", exist_ok=True)
os.makedirs("outputs/tables", exist_ok=True)

print("Project directories ready.")

Project directories ready.


## 2. Loading the Dataset

This project uses the Heart Disease dataset containing clinical attributes associated with cardiovascular health.

The objective is to classify whether a patient shows the presence or absence of heart disease using a Support Vector Machine classifier.

In [5]:
import pandas as pd

print(pd.__version__)

3.0.6


In [7]:
# Load the Cleveland Heart Disease dataset from the official UCI repository

dataset_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data"

column_names = [
    "age", "sex", "cp", "trestbps", "chol", "fbs",
    "restecg", "thalach", "exang", "oldpeak",
    "slope", "ca", "thal", "target"
]

df = pd.read_csv(
    dataset_url,
    names=column_names,
    na_values="?"
)

df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63.0,1.0,1.0,145.0,233.0,1.0,2.0,150.0,0.0,2.3,3.0,0.0,6.0,0
1,67.0,1.0,4.0,160.0,286.0,0.0,2.0,108.0,1.0,1.5,2.0,3.0,3.0,2
2,67.0,1.0,4.0,120.0,229.0,0.0,2.0,129.0,1.0,2.6,2.0,2.0,7.0,1
3,37.0,1.0,3.0,130.0,250.0,0.0,0.0,187.0,0.0,3.5,3.0,0.0,3.0,0
4,41.0,0.0,2.0,130.0,204.0,0.0,2.0,172.0,0.0,1.4,1.0,0.0,3.0,0


In [8]:
# Display the dataset dimensions and column names

print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

print("\nFeatures:")
print(df.columns.tolist())

Rows: 303
Columns: 14

Features:
['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target']


In [9]:
# Save the raw dataset locally and export a preview table

df.to_csv("data/raw/heart_disease_cleveland.csv", index=False)

df.head(10).to_csv(
    "outputs/tables/dataset_preview.csv",
    index=False
)

print("Dataset saved successfully.")

Dataset saved successfully.


### Dataset Loading Insights

- The Cleveland Heart Disease dataset contains 303 patient records.
- It consists of 13 clinical predictor variables and one diagnosis target variable.
- Missing values represented by `?` in the original data were converted to standard missing values during loading.
- The original target ranges from 0 to 4, where 0 represents absence of heart disease and values 1–4 indicate its presence.
- A local copy of the dataset has been stored for reproducibility.